### PR TITLE
Add SASL_SCRAM authentication mechanism on kafka exporter

### LINKED ISSUE #2252: Add SASL/SCRAM authentication to Kafka receiver and exporter
**Is your feature request related to a problem? Please describe.**
I would like to use SASL/SCRAM authentication mechanism on kafka receiver and exporter.
The current authentication mechanisms implemented on collector are not supported by AWS MSK

**Describe the solution you'd like**

SASL/SCRAM authentication mechanism is already supported by [sarama](https://github.com/Shopify/sarama/tree/master/examples/sasl_scram_client), the library used by collector.
We could implement this on [authentication.go](https://github.com/open-telemetry/opentelemetry-collector/blob/master/exporter/kafkaexporter/authentication.go) using the same library, adding a SCRAMConfig struct and a new function to configure SCRAM.

**Describe alternatives you've considered**
None.
**Additional context**
None.

### PR DESCRIPTION
**Pull request by 4lt1**
**Description:** 
This PR adds SASL/SCRAM authentication to kafka exporter and receiver.

**Link to tracking Issue:** 
Resolves #2252

**Documentation:**
Added to readme

### CODE REVIEW

**Comment by tigrannajaryan** [comment]
@jpkrohling can you review this please?

**Comment by jpkrohling** [comment]
Sorry, was on a long break. Will review this one soon.

File: exporter/kafkaexporter/authentication.go [hunk 1]
```diff
@@ -68,6 +84,29 @@ func configurePlaintext(config PlainTextConfig, saramaConfig *sarama.Config) {
 	saramaConfig.Net.SASL.Password = config.Password
 }
 
+func configureSCRAM(config SCRAMConfig, saramaConfig *sarama.Config) error {
+	saramaConfig.Net.SASL.Enable = true
+	saramaConfig.Net.SASL.User = config.Username
+	saramaConfig.Net.SASL.Password = config.Password
+	if config.Algorithm != "sha256" && config.Algorithm != "sha512" {
```
Comment (jpkrohling) [inline exporter/kafkaexporter/authentication.go:91]: I feel like a switch statement works better here.

File: exporter/kafkaexporter/scram_client.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kafkaexporter
+
+import (
+	"github.com/xdg/scram"
+)
+
+// XDGSCRAMClient defines XDGSCRAMClient
```
Comment (jpkrohling) [inline exporter/kafkaexporter/scram_client.go:21]: The godoc can be improved ;-)

File: exporter/kafkaexporter/scram_client.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	x.Client, err = x.HashGeneratorFcn.NewClient(userName, password, authzID)
+	if err != nil {
+		return err
+	}
+	x.ClientConversation = x.Client.NewConversation()
+	return nil
+}
+
+func (x *XDGSCRAMClient) Step(challenge string) (response string, err error) {
+	response, err = x.ClientConversation.Step(challenge)
```
Comment (jpkrohling) [inline exporter/kafkaexporter/scram_client.go:38]: You can just return from here, can't you?

File: exporter/kafkaexporter/scram_client.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kafkaexporter
+
+import (
+	"github.com/xdg/scram"
```
Comment (jpkrohling) [inline exporter/kafkaexporter/scram_client.go:18]: github.com/xdg/scram redirects to github.com/xdg-go/scram

**Review by jpkrohling — COMMENTED** [review]
This is missing quite a few tests, but looks like it's a good start.

File: exporter/kafkaexporter/authentication.go (hunk 1 above)
Comment (4lt1) [reply exporter/kafkaexporter/authentication.go:91]: Yeah! I totally agree.
I'll change it.

**Comment by 4lt1** [comment]
> This is missing quite a few tests, but looks like it's a good start.
 Thanks for your comments @jpkrohling, I'll write the tests this week.

File: exporter/kafkaexporter/authentication.go [hunk 5]
```diff
@@ -35,6 +38,13 @@ type PlainTextConfig struct {
 	Password string `mapstructure:"password"`
 }
 
+// SCRAMConfig defines plaintext authentication.
```
Comment (jpkrohling) [inline exporter/kafkaexporter/authentication.go:41]:
```suggestion
// SCRAMConfig defines the configuration for the SCRAM authentication.
```

File: exporter/kafkaexporter/authentication.go [hunk 6]
```diff
@@ -68,6 +84,25 @@ func configurePlaintext(config PlainTextConfig, saramaConfig *sarama.Config) {
...
+
+	switch config.Algorithm {
+	case "sha512":
+		saramaConfig.Net.SASL.SCRAMClientGeneratorFunc = func() sarama.SCRAMClient { return &XDGSCRAMClient{HashGeneratorFcn: sha512.New} }
+		saramaConfig.Net.SASL.Mechanism = sarama.SASLTypeSCRAMSHA512
+	case "sha256":
+		saramaConfig.Net.SASL.SCRAMClientGeneratorFunc = func() sarama.SCRAMClient { return &XDGSCRAMClient{HashGeneratorFcn: sha256.New} }
+		saramaConfig.Net.SASL.Mechanism = sarama.SASLTypeSCRAMSHA256
+	default:
+		return fmt.Errorf("invalid SHA algorithm \"%s\": can be either \"sha256\" or \"sha512\"", config.Algorithm)
```
Comment (jpkrohling) [inline exporter/kafkaexporter/authentication.go:100]:
```suggestion
		return fmt.Errorf("invalid SHA algorithm %q: can be either \"sha256\" or \"sha512\"", config.Algorithm)
```

File: exporter/kafkaexporter/scram_client.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+// limitations under the License.
+
+package kafkaexporter
+
+import (
+	"github.com/xdg-go/scram"
+)
+
+// XDGSCRAMClient uses xdg-go scram to authentication conversation
+type XDGSCRAMClient struct {
```
Comment (jpkrohling) [inline exporter/kafkaexporter/scram_client.go:22]: You should probably do a type assert here, ensuring that this is a `sarama.SCRAMClient`

**Review by jpkrohling — COMMENTED** [review]
This is still missing a test and the build seems to be failing, but other than that, looks sane.

**Comment by 4lt1** [comment]
Hi @jpkrohling, sorry about the delay.

I've written some unit tests and changed the Authentication type from SCRAM to SASL.
In my point of view, to use sasl.mechanism = "SCRAM-SHA-256| SCRAM-SHA-512|PLAIN" is more clear and compliant with what is described by [SASL Mechanisms](https://www.iana.org/assignments/sasl-mechanisms/sasl-mechanisms.xhtml) and very similar with [kafka](https://docs.confluent.io/3.0.0/kafka/sasl.html) and this merged  on  [jaeger](https://github.com/jaegertracing/jaeger/pull/2724) project.

Maybe we could deprecate [PlainText](https://github.com/open-telemetry/opentelemetry-collector/blob/1ed9d6f23b533a628815ecdc81933ec5c365f33e/exporter/kafkaexporter/authentication.go#L29) to use sasl.mechanism = "PLAIN" in the future.

File: exporter/kafkaexporter/README.md [hunk 8]
```diff
@@ -19,6 +19,10 @@ The following settings can be optionally configured:
   - `plain_text`
     - `username`: The username to use.
     - `password`: The password to use
+  - `sasl`
+    - `username`: The username to use.
+    - `password`: The password to use
+    - `mechanism`: The sasl mechanism to use (SCRAM-SHA-256 or SCRAM-SHA-512)
```
Comment (jpkrohling) [inline exporter/kafkaexporter/README.md:25]: And `PLAIN` as discussed in the comments

File: exporter/kafkaexporter/authentication_test.go [hunk 9]
```diff
@@ -86,6 +123,8 @@ func TestAuthentication(t *testing.T) {
 				require.Error(t, err)
 				assert.Contains(t, err.Error(), test.err)
 			} else {
+				// equaequalizes SCRAMClientGeneratorFunc to do assertion with the same reference.
```
Comment (jpkrohling) [inline exporter/kafkaexporter/authentication_test.go:126]:
```suggestion
				// equalizes SCRAMClientGeneratorFunc to do assertion with the same reference.
```

**Review by jpkrohling — COMMENTED** [review]
Looks good then, it's only missing the PLAIN in one place.

**Review by jpkrohling — APPROVED** [review]
I haven't tried this PR myself as it's not that easy for me to get a properly configured Kafka instance, but once @4lt1 confirms that it was manually tested, this is ready to be merged.

**Comment by jpkrohling** [comment]
@open-telemetry/collector-maintainers, would one of you please review/merge this one?

**Comment by jpkrohling** [comment]
@open-telemetry/collector-maintainers, would one of you please review/merge this one?

File: exporter/kafkaexporter/authentication.go [hunk 10]
```diff
@@ -35,6 +38,13 @@ type PlainTextConfig struct {
 	Password string `mapstructure:"password"`
 }
 
+// SASLConfig defines the configuration for the SASL authentication.
+type SASLConfig struct {
+	Username  string `mapstructure:"username"`
+	Password  string `mapstructure:"password"`
+	Mechanism string `mapstructure:"mechanism"`
```
Comment (bogdandrutu) [inline exporter/kafkaexporter/authentication.go:45]: Please add a comment about expected values

File: go.mod [hunk 11]
```diff
@@ -48,6 +48,7 @@ require (
 	github.com/stretchr/testify v1.7.0
 	github.com/tinylib/msgp v1.1.5
 	github.com/uber/jaeger-lib v2.4.0+incompatible
+	github.com/xdg-go/scram v0.0.0-20180814205039-7eeb5667e42c
```
Comment (bogdandrutu) [inline go.mod:51]: How ok are we to depend on a very old, not actively maintained package?

Does not even have go[mod/sum].

Comment (jpkrohling) [reply go.mod:51]: Honestly? It doesn't worry me that much, as the code is fairly simple and based on a couple of RFCs. Once the standard is implemented, there's not much else to do. I do have thought about the responsiveness of the author in that repository (see the issue opened by @yurishkuro) but I don't see a reason for implementing this ourselves at this moment.

**Comment by bogdandrutu** [comment]
@4lt1 do you mind creating a new PR (just clone the branch and open a new PR) to make circleCI trigger to run the tests? Thank you, and sorry for this

**Comment by 4lt1** [comment]
> @4lt1 do you mind creating a new PR (just clone the branch and open a new PR) to make circleCI trigger to run the tests? Thank you, and sorry for this

No problem at all @bogdandrutu, the new PR is https://github.com/open-telemetry/opentelemetry-collector/pull/2503

**Comment by jpkrohling** [comment]
I'm closing this in favor of #2503

**Pull request closed without merging**